xmsannotator/R/compute_scores.R

```diff
@@ -1,35 +1,157 @@
+    rt_cluster == query$rt_cluster,
+    mean_intensity / query$mean_intensity <= abundance_ratio + abundance_ratio * intensity_deviation_tolerance,
     near(rt, query$rt, rt_tolerance),
-    near(mass_defect, query$mass_defect, mass_defect_tolerance),
-    between(abs(isotope_number), 1, max_isp)
+    near(mass_defect, query$mass_defect, mass_defect_tolerance)
   )
-  isotopes <- select(isotopes, -isotope_number)
 }
 
+#' Match prefiltered isotopes by comparing their relative intensity to their relative natural abundance.
+#' The relative values are computed as a ratio of an isotopic peak intensity to the intensity of its monoisotopic peak,
+#' respectively as a ratio of a natural abundance of the isotope to the abundance of the most abundant isotope.
+#'
+#' @param query A single annotated peak.
+#' @param isotopes Isotope candidates for the annotated molecule. The candidates are identified in [compute_isotopes].
+#' @param pattern Isotopic pattern of a given molecule as produced by [compute_isotopic_pattern].
+#' @param intensity_deviation_tolerance A numeric threshold by which an intensity ratio of two isotopic peaks may differ
+#'  from their actual abundance ratio.
+#'
+#' @return A table of matched isotopes.
+#'
+#' @import dplyr
+match_isotopes_by_intensity <- function(query,
+                                        isotopes,
+                                        pattern,
+                                        intensity_deviation_tolerance) {
+  isotopes <- mutate(isotopes,
+                     relative_intensity = mean_intensity / query$mean_intensity)
+  isotopes <- left_join(isotopes,
+                        select(pattern, mass_number_difference, abund),
+                        by = "mass_number_difference")
+  isotopes <- filter(isotopes,
+                     near(relative_intensity, abund, relative_intensity * intensity_deviation_tolerance))
+  isotopes <- select(isotopes, -c(abund, relative_intensity))
+}
+
+#' Find all possible isotopes of a given annotated peak by filtering peaks based on several criteria
+#' (see [filter_isotopes]) and matching the intensities with natural abundances of the isotopes.
+#'
+#' @param ... A single isotope passed as a list of columns. See [purrr::map()] for more details.
+#' @param intensity_deviation_tolerance A numeric threshold by which an intensity ratio of two isotopic peaks may differ
+#'  from their actual abundance ratio.
+#' @param peaks A peak table containing a peak identifier (unique number), mean intensity, module, and rt cluster
+#'  of each identified peak.
+#' @param mass_defect_tolerance A number. Maximum difference in mass defect between two peaks of the same compound.
+#' @param rt_tolerance A number. Maximum rt difference for two peaks of the same substance.
+#'
+#' @return A table with peaks that have been identified as isotopes of a given molecule from the annotation table.
+#'
+#' @import dplyr
+#' @importFrom rlang .data
+compute_isotopes <- function(...,

```

Comment:
To be coherent with the formulation above, this shoulb be renamed to 'detect_isotopic_peaks' or so, since we detect the isotopic peaks for a given query annotation, right?